src/ray/gcs/gcs_server/ray_sync.h

```diff
@@ -0,0 +1,117 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+
+#include <memory>
+#include <type_traits>
+#include "ray/gcs/gcs_server/grpc_based_resource_broadcaster.h"
+#include "ray/gcs/gcs_server/gcs_resource_report_poller.h"
+#include "ray/common/asio/periodical_runner.h"
+
+namespace ray {
+namespace sync {
+
+class RaySync {
+ public:
+  RaySync(instrumented_io_context& main_thread,
+          std::unique_ptr<::ray::gcs::GrpcBasedResourceBroadcaster> braodcaster,
+          std::unique_ptr<::ray::gcs::GcsResourceReportPoller> poller)
+      : 
+      ticker_(main_thread),
+      broadcaster_(std::move(braodcaster)), 
+      poller_(std::move(poller)) {}
+
+  void Start() {
```

Comment:
I like the API! It is also pretty easy to understand things (it is a bit sad poller has its own loop instead of directly calling poller->Poll() here, but it seems to be difficult). 

Is there any reason why we don't run poller in a separate thread too btw? I feel like this should give clear benefit in the large scale cluster, and seems to be low hanging fruit

Reply:
poller is running in a separate thread right now. Btw, this is just the first step. We'll change it eventually to a push-based model like this prototype (https://github.com/ray-project/ray/pull/22073).

Reply:
Hmm you mean we will go back to push based report from pull based?

Reply:
Right now, I think that's still arguable. So the plan is 1) make it lightweight (we need it anyway); 2) benchmark both to make the decision.  The pros are that with streaming, it seems way easier to do it in a push way.  Anyway, that might be very late.

Reply:
I see. This sounds reasonable. For other parts we can definitely discuss it again once it is ready to experiment